File: procedural/queens.rb
#!/usr/bin/env ruby

def nqueens(size)
    row = Array.new(size, true)
    up = Array.new(2*size-1, true)
    down = Array.new(2*size-1, true)
    board = Array.new(size)
    queen = lambda do |n|
        if n >= size then
            puts board.join(' ')
        else
            size.times do |i|
                if row[i] && up[n+i] && down[n-i] then
                   row[i] = up[n+i] = down[n-i] = false
                   board[n] = i+1
                   queen.call(n+1)
                   row[i] = up[n+i] = down[n-i] = true
                end
            end
        end
    end
    queen.call(0)
end

nqueens(ARGV[0].to_i)

File: procedural/queens.kt
fun main(args: Array<String>) { 
    fun nqueens(size: Int) {
        val row   = BooleanArray(size){true}
        val up    = BooleanArray(2*size-1){true}
        val down  = BooleanArray(2*size-1){true}
        val board = IntArray(size)
        fun queen(n: Int) { 
            if (n >= size) { 
                for (i in 0..size-2) { 
                    print("${board[i]} ")
                } 
                println("${board[size-1]}")
            } else { 
                for (i in 0..size-1) { 
                    if (row[i] && up[n+i] && down[n-i+size-1]) { 
                        row[i]           = false
                        up[n+i]          = false
                        down[n-i+size-1] = false
                        board[n] = i+1
                        queen(n+1)
                        row[i]           = true
                        up[n+i]          = true
                        down[n-i+size-1] = true
                    } 
                } 
            } 
        }
        queen(0)
    }
    nqueens(args[0].toInt())
}
